x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/GetCcrRestoreFileChunkAction.java

```diff
@@ -0,0 +1,130 @@
+
+import java.io.IOException;
+
+public class GetCcrRestoreFileChunkAction extends Action<GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse> {
+
+    public static final GetCcrRestoreFileChunkAction INSTANCE = new GetCcrRestoreFileChunkAction();
+    public static final String NAME = "internal:admin/ccr/restore/file_chunk/get";
+
+    private GetCcrRestoreFileChunkAction() {
+        super(NAME);
+    }
+
+    @Override
+    public GetCcrRestoreFileChunkResponse newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Writeable.Reader<GetCcrRestoreFileChunkResponse> getResponseReader() {
+        return GetCcrRestoreFileChunkResponse::new;
+    }
+
+
+    public static class TransportGetCcrRestoreFileChunkAction
+        extends HandledTransportAction<GetCcrRestoreFileChunkRequest, GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse> {
+
+        private final CcrRestoreSourceService restoreSourceService;
+        private final ThreadPool threadPool;
+        private final BigArrays bigArrays;
+
+        @Inject
+        public TransportGetCcrRestoreFileChunkAction(BigArrays bigArrays, TransportService transportService, ActionFilters actionFilters,
+                                                     CcrRestoreSourceService restoreSourceService) {
+            super(NAME, transportService, actionFilters, GetCcrRestoreFileChunkRequest::new);
+            TransportActionProxy.registerProxyAction(transportService, NAME, GetCcrRestoreFileChunkResponse::new);
+            this.threadPool = transportService.getThreadPool();
+            this.restoreSourceService = restoreSourceService;
+            this.bigArrays = bigArrays;
+        }
+
+        @Override
+        protected void doExecute(Task task, GetCcrRestoreFileChunkRequest request,
+                                 ActionListener<GetCcrRestoreFileChunkResponse> listener) {
+            threadPool.generic().execute(new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+
+                @Override
+                protected void doRun() throws Exception {
+                    int bytesRequested = request.getSize();
+                    long fileOffset = request.getOffset();
+                    ByteArray array = bigArrays.newByteArray(bytesRequested, false);
+                    String fileName = request.getFileName();
+                    String sessionUUID = request.getSessionUUID();
+                    // This is currently safe to do because calling `onResponse` will serialize the bytes to the network layer data
+                    // structure on the same thread. So the bytes will be copied before the reference is released.
+                    try (ReleasablePagedBytesReference reference = new ReleasablePagedBytesReference(array, bytesRequested, array)) {
```

Comment:
I looked at this again and I think this API is not in a good shape. We allow to read from arbitrary offsets here, concurrently. I don't think we should do that, at least not with this impl. On [line 93](https://github.com/elastic/elasticsearch/pull/37130/files#diff-0c95985a3259313c3536dfa937926eb5R93) we iterate over one or more references which assume consecutive reads. That can't be guaranteed even not with the synchronization in `Reader#readFileBytes`. There are more issues in there with regards to IndexInput cloning. That said, we either should use a random access API but then also make use of it, in that case we shouldn't use IndexInput under the hood but rather a FileChannel that offers us the right primitives. If we do that we should make sure we align the bytesRequested with the page size in BigArrays. 

The other option is to keep it simple and add a read-sequential API that doesn't allow specifying the offset. Here we should fail if there are more than one thread accessing the same <sessionID, fileName> tuple concurrently. We can, in addition allow the client to specify a token that it get's back from each read to ensure we are still in the same state. The current model seems dangerous to me, I'd appreciate a simpler model also we can still fetch multiple files concurrently instead of having concurrency on the chunk level at least API wise.